xbmc/addons/binary/BinaryAddonManager.cpp

```diff
@@ -0,0 +1,303 @@
+/*
+ *      Copyright (C) 2015-2016 Team KODI
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with KODI; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "BinaryAddonManager.h"
+//#include "binary-executable/BinaryAddon.h"
+//#include "binary-executable/tools.h"
+
+#include "Application.h"
+#include "utils/log.h"
+
+#include <netinet/in.h>
+
+#define LOCK_AND_COPY(type, dest, src) \
+  if (!m_bInitialized) return; \
+  CSingleLock lock(src); \
+  src.hadSomethingRemoved = false; \
+  type dest; \
+  dest = src
+
+#define CHECK_FOR_ENTRY(l,v) \
+  (l.hadSomethingRemoved ? (std::find(l.begin(),l.end(),v) != l.end()) : true)
+
+namespace ADDON
+{
+
+//unsigned int CBinaryAddonManager::m_IdCnt = 0;
+
+CBinaryAddonManager::CBinaryAddonManager()
+ : CThread("Binary add-on manager"),
+   m_bInitialized(false)
+{
+}
+
+CBinaryAddonManager::~CBinaryAddonManager()
+{
+//  m_Status.Shutdown();
+}
+
+CBinaryAddonManager &CBinaryAddonManager::GetInstance()
```

Comment:
TBH I'm not sure if having all kinds of stuff in `CApplication` and then use it through the `g_application` global is any better than a singleton. Yes you have more control over initialization and deinitialization but you can get the same by explicitly initializing and uninitializing a singleton. Actually I prefer to have a direct dependency on the class I'm using instead of everything depending on `CApplication` which in turn depends on all kinds of other stuff. That way the dependencies are less obvious and changing `CApplication` will result in having to re-compile a lot of other theoretically unrelated code.